### ISSUE TITLE
Add support for E-Mu EIV devices

### ISSUE THREAD

**Issue by dagargo** [dagargo/emu3fs#3]
I'm gonna need some ISO images to implement this. If anyone is interested in this issue, provide me with some. Thanks!

**Comment by dromer** [comment]
I formatted a zip drive on my Emulator-IV that I'm now dd-ing an iso from.
What's your email so I can send you a download link? (don't want to put it here as it might overload my server hehe)

The iso is completely empty btw, just formatted the drive and that's it.

**Comment by dromer** [comment]
Also, I did try to simply `mount -t emu3 /dev/sdh /mnt/emu` and the kernel module did not complain .. so maybe it already works?™ :D

**Comment by dromer** [comment]
I created an empty bank on the disk and after mounting I can see `-rw-rw-rw- 1 dreamer dreamer 11K May 14 14:11 'Untitled Bank'`

However emu3bm says the format is not supported. Seems that the FS is quite compatible, but work on emu3bm is needed to be able to read the data.

**Comment by raybellis** [comment]
E4 banks look very different from ESI / E-III banks - they use RIFF style headers.

Also, unless you create a subfolder (or two) the empty EIV "ISO" will likely look no different to an E-III disk, since the only real difference is in how those subfolders are listed in the directory structure.  An ideal ISO for analysis would need a few banks in the top folder, and at least a couple of subfolders with banks in too.

**Comment by dromer** [comment]
Good to know @raybellis I'll go try and add some more stuff to this disk.

**Comment by dagargo** [comment]
Thanks to you both.

@dromer, my email is now made public in my profile.

**Comment by dromer** [comment]
@dagargo email sent :)

**Comment by dagargo** [comment]
Email received. Thanks!

At first glance, I have 2 interpretations for what I see.

1. There are 3 files named `Untitled Bank` and one folder named `New Folder` which contains another bank equally named.
2. There are 4 files named `Untitled Bank`.

What did you send? First interpretation? Anything else?

BTW, I forgot to mention that if you truncate the file at a few MB we will have more than enough.

Also, regarding the EIV banks comment above, it is really unlikely I will try to research on that because it's a lot of work, mostly useless and I have no EIV devices.

**Comment by dromer** [comment]
hmm, I thought the first 3 banks should be in one folder, and then the other bank in the second folder, but you're saying that you only see a single folder on here?

Being able to edit EIV banks is only useless if you don't have an EIV device, but for others it would be nice ;)

However I think the EIV can also read E3 disks, so worst case I stick all my sample work to that and exchange the same format between the ESI-4000 and EIV that way.

**Comment by dagargo** [comment]
> Being able to edit EIV banks is only useless if you don't have an EIV device, but for others it would be nice ;)

Touché.

What I see in the image is this.
```
00001000  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00001010  00 80 09 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00001020  4e 65 77 20 46 6f 6c 64  65 72 20 20 20 20 20 00  |New Folder     .|
00001030  00 40 0a 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
[...]
00001200  55 6e 74 69 74 6c 65 64  20 42 61 6e 6b 20 20 20  |Untitled Bank   |
00001210  00 00 01 00 01 00 15 00  f2 01 81 00 45 34 42 30  |............E4B0|
00001220  55 6e 74 69 74 6c 65 64  20 42 61 6e 6b 20 20 20  |Untitled Bank   |
00001230  00 01 02 00 01 00 15 00  f2 01 81 00 45 34 42 30  |............E4B0|
00001240  55 6e 74 69 74 6c 65 64  20 42 61 6e 6b 20 20 20  |Untitled Bank   |
00001250  00 02 03 00 01 00 15 00  f2 01 81 00 45 34 42 30  |............E4B0|
[...]
00001400  55 6e 74 69 74 6c 65 64  20 42 61 6e 6b 20 20 20  |Untitled Bank   |
00001410  00 00 04 00 01 00 15 00  f2 01 81 00 45 34 42 30  |............E4B0|
```

And, once mounted, I see this.
```
$ ls -l /media/emu3/
total 42
-rw-rw-rw- 1 user user 10738 may 15 14:07 'Untitled Bank'
-rw-rw-rw- 1 user user 10738 may 15 14:07 'Untitled Bank'
-rw-rw-rw- 1 user user 10738 may 15 14:07 'Untitled Bank'
-rw-rw-rw- 1 user user 10738 may 15 14:07 'Untitled Bank'
```

What is interesting is that looks like EIV format is backward compatible with EIII and that's why we can see the 4 banks there. Most likely, the folder support was already there but not supported by the hardware. But perhaps they found a way to implement this while keeping it compatible. We'll never know.

Tipically, EIII images have the same section and then the banks.
```
00000c00  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00000c10  00 40 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
[...]
00001200  30 6e 30 69 74 6c 65 64  20 42 61 6e 6b 20 20 20  |Untitled Bank   |
00001210  00 00 01 00 01 00 1f 00  39 00 81 00 00 00 00 00  |........9.......|
```

I was totally clueless about this section but perhaps it makes sense now.

BTW, In this example, the addresses are the same but IIRC this depends on the format so addresses may vary from disk to disk.

Could you confirm that the first folder is called `Default Folder` and the send is `New Folder`?

**Comment by dromer** [comment]
@dagargo yes if I go to `Disk > Iomega Zip > Browse > Folders` I see `Default Folder` and `New Folder`

`Default Folder` contains 3 banks, and `New Folder` contains 1.

Maybe I should've given everything non-default names to make it easier to guess what the setup "should" be :)

**Comment by raybellis** [comment]
It looks like the byte (or 2 byte big-endian short) at offset 0x12 in each of the two directory entries is an offset to the two (non-contiguous) blocks of file entries (512 bytes per block).

i.e. 0x0009 * 0x0200 = 0x00001200
     0x000a * 0x0200 = 0x00001400

**Comment by dagargo** [comment]
Well spotted, @raybellis.

I was researching a bit and I found this piece of information in the EVI operation manual found [here](http://www.theemus.com/documentation/eiv/EIV_Operation_Manual.pdf).

> A folder can contain up to 100 banks (0-99). You can use folders to organize your sound banks or you might want to include all the banks used for a particular project in a folder.The number of folders in a disk depends on the size of the disk. As an example, a one gigabyte hard disk can hold 96 folders.

We already know these facts.
* EIII devices are able to show the first 100 banks in the whole disk and probably will write to the `Default Folder` and are able to ignore the other folders somehow.
* EIV use that folder as any other folder and not as the _root_ folder. Also, any folder can store up to 100 banks. 

So, before looking into the code I think we have a problem. If we make the EIII FS fully compatible with EIV, how do we know that folders can be created (EIV) and if there is a root folder (EIII)? I don't think it's a good idea to support folders for EIII disks because it will be confusing. Perhaps, we could mount devices with an option or perhaps there is a way to detect the type of device that created the FS. Let's research on this.

**Comment by dagargo** [comment]
I forgot to mention that when mounting the device some description of what I called _geometry_ of the filesystem is written by the kernel. If you're curious, take a look at the code for the actual meaning.

**Comment by dagargo** [comment]
Just wanted to say that I had an ESI 2000 formatted disk with more than 40 banks like this.

```
00000c00  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00000c10  00 40 09 00 0a 00 0b 00  0c 00 0d 00 0e 00 0f 00  |.@..............|
```

And after I changed the line we talked about before like this, then only 16 banks were detected by the ESI 2000.
```
00000c00  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00000c10  00 40 09 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
```

As there are 7 shorts available, which are the block addresses, and every block can store up to 16 banks that would mean that a folder can store up to 112 (7*16) banks and this is probably why they limited to 100 (0-99).

Also, I can confirm that old banks created for the Emulator III, do not have this `Default folder` so they just used a fixed position for the 7 blocks (always 100 banks).

Looks like all this confirms @raybellis suspicion and explains everything.

But I haven't found anything yet that would let us detect automatically the FS.

**Comment by dagargo** [comment]
Just a small update.

For every image disk I have, there is a folder. So what I said in the previous comment is **not** true.
Of course, in the EIV format, there could be more. 

```
$ for f in *.iso; do echo $f; hexdump -C -v $f | head -n 3; b=0x$(hexdump -C -v $f | head -n 1 | awk '{print $10}'); ab=$(printf '%x' $((b * 512))); xa=0000$ab; hexdump -C -v $f | head -n 1000 | grep $xa -A 1; done
clean-zip.iso
00000000  45 4d 55 33 ff ff 02 00  08 00 00 00 01 00 00 00  |EMU3............|
00000010  09 00 00 00 6f 00 00 00  02 00 00 00 06 00 00 00  |....o...........|
00000020  78 00 00 00 ff 05 00 00  01 01 00 00 00 08 00 00  |x...............|
00001000  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00001010  00 40 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
emu_iv_format.iso
00000000  45 4d 55 33 ff ff 02 00  08 00 00 00 01 00 00 00  |EMU3............|
00000010  09 00 00 00 6f 00 00 00  02 00 00 00 06 00 00 00  |....o...........|
00000020  78 00 00 00 ff 05 00 00  01 01 00 02 ff ff 00 00  |x...............|
00001000  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00001010  00 80 09 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
emu_std_v10.iso
00000000  45 4d 55 33 00 cd 08 00  07 00 00 00 02 00 00 00  |EMU3............|
00000010  09 00 00 00 65 00 00 00  02 00 00 00 05 00 00 00  |....e...........|
00000020  6e 00 00 00 66 04 00 00  03 01 e6 cd 08 00 00 00  |n...f...........|
00000e00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000e10  00 40 09 00 0a 00 0b 00  0c 00 0d 00 0e 00 0f 00  |.@..............|
emu_std_v12-1.iso
00000000  45 4d 55 33 e6 cd 08 00  07 00 00 00 02 00 00 00  |EMU3............|
00000010  09 00 00 00 65 00 00 00  02 00 00 00 05 00 00 00  |....e...........|
00000020  6e 00 00 00 66 04 00 00  03 01 00 08 cd e6 00 00  |n...f...........|
00000e00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000e10  00 40 09 00 0a 00 0b 00  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v12-2.iso
00000000  45 4d 55 33 00 55 19 00  09 00 00 00 06 00 00 00  |EMU3.U..........|
00000010  0f 00 00 00 c0 00 00 00  02 00 00 00 07 00 00 00  |................|
00000020  cf 00 00 00 55 06 00 00  04 01 8a 55 19 00 00 00  |....U......U....|
00001200  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00001210  00 40 0f 00 10 00 ff ff  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v13.iso
00000000  45 4d 55 33 00 55 19 00  09 00 00 00 06 00 00 00  |EMU3.U..........|
00000010  0f 00 00 00 c0 00 00 00  02 00 00 00 07 00 00 00  |................|
00000020  cf 00 00 00 55 06 00 00  04 01 8a 55 19 00 00 00  |....U......U....|
00001200  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00001210  00 40 09 00 10 00 11 00  12 00 ff ff ff ff ff ff  |.@..............|
emu_std_v16.iso
00000000  45 4d 55 33 00 55 19 00  09 00 00 00 06 00 00 00  |EMU3.U..........|
00000010  0f 00 00 00 c0 00 00 00  02 00 00 00 07 00 00 00  |................|
00000020  cf 00 00 00 55 06 00 00  04 01 8a 55 19 00 00 00  |....U......U....|
00001200  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00001210  00 40 09 00 10 00 11 00  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v1.iso
00000000  45 4d 55 33 00 f4 08 00  07 00 00 00 02 00 00 00  |EMU3............|
00000010  09 00 00 00 66 00 00 00  02 00 00 00 05 00 00 00  |....f...........|
00000020  6f 00 00 00 79 04 00 00  03 01 0b f4 08 00 00 00  |o...y...........|
00000e00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000e10  00 40 09 00 0a 00 0b 00  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v2.iso
00000000  45 4d 55 33 00 f4 08 00  07 00 00 00 02 00 00 00  |EMU3............|
00000010  09 00 00 00 66 00 00 00  02 00 00 00 05 00 00 00  |....f...........|
00000020  6f 00 00 00 79 04 00 00  03 01 0b f4 08 00 00 00  |o...y...........|
00000e00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000e10  00 40 09 00 0a 00 0b 00  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v3.iso
00000000  45 4d 55 33 00 ad 0f 00  06 00 00 00 04 00 00 00  |EMU3............|
00000010  0a 00 00 00 7b 00 00 00  02 00 00 00 04 00 00 00  |....{...........|
00000020  85 00 00 00 eb 03 00 00  04 01 db ad 0f 00 00 00  |................|
00000c00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000c10  00 40 09 00 0b 00 0c 00  0d 00 ff ff ff ff ff ff  |.@..............|
emu_std_v4.iso
00000000  45 4d 55 33 00 ad 0f 00  06 00 00 00 04 00 00 00  |EMU3............|
00000010  0a 00 00 00 7b 00 00 00  02 00 00 00 04 00 00 00  |....{...........|
00000020  85 00 00 00 eb 03 00 00  04 01 db ad 0f 00 00 00  |................|
00000c00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000c10  00 40 09 00 0b 00 ff ff  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v5.iso
00000000  45 4d 55 33 00 ad 0f 00  06 00 00 00 04 00 00 00  |EMU3............|
00000010  0a 00 00 00 7b 00 00 00  02 00 00 00 04 00 00 00  |....{...........|
00000020  85 00 00 00 eb 03 00 00  04 01 db ad 0f 00 00 00  |................|
00000c00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000c10  00 40 09 00 0a 00 0b 00  0c 00 0d 00 0e 00 0f 00  |.@..............|
emu_std_v6.iso
00000000  45 4d 55 33 00 ad 0f 00  06 00 00 00 04 00 00 00  |EMU3............|
00000010  0a 00 00 00 7b 00 00 00  02 00 00 00 04 00 00 00  |....{...........|
00000020  85 00 00 00 eb 03 00 00  04 01 db ad 0f 00 00 00  |................|
00000c00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000c10  00 40 09 00 0b 00 ff ff  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v7.iso
00000000  45 4d 55 33 00 ad 0f 00  06 00 00 00 04 00 00 00  |EMU3............|
00000010  0a 00 00 00 7b 00 00 00  02 00 00 00 04 00 00 00  |....{...........|
00000020  85 00 00 00 eb 03 00 00  04 01 db ad 0f 00 00 00  |................|
00000c00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000c10  00 40 09 00 0b 00 ff ff  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_v8.iso
00000000  45 4d 55 33 00 ad 0f 00  06 00 00 00 04 00 00 00  |EMU3............|
00000010  0a 00 00 00 7b 00 00 00  02 00 00 00 04 00 00 00  |....{...........|
00000020  85 00 00 00 eb 03 00 00  04 01 db ad 0f 00 00 00  |................|
00000c00  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00000c10  00 40 09 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
emu_std_vol16.iso01.iso
00000000  45 4d 55 33 00 55 19 00  09 00 00 00 06 00 00 00  |EMU3.U..........|
00000010  0f 00 00 00 c0 00 00 00  02 00 00 00 07 00 00 00  |................|
00000020  cf 00 00 00 55 06 00 00  04 01 8a 55 19 00 00 00  |....U......U....|
00001200  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00001210  00 40 0f 00 10 00 11 00  ff ff ff ff ff ff ff ff  |.@..............|
formula_4000.iso
00000000  45 4d 55 33 29 e8 1f 00  06 00 00 00 06 00 00 00  |EMU3)...........|
00000010  0c 00 00 00 9a 00 00 00  02 00 00 00 04 00 00 00  |................|
00000020  a6 00 00 00 fc 03 00 00  05 01 00 00 00 08 00 00  |................|
00000c00  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00000c10  00 40 09 00 ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
test-image.iso
00000000  45 4d 55 33 ff 1e 0f 00  06 00 00 00 03 00 00 00  |EMU3............|
00000010  09 00 00 00 6a 00 00 00  02 00 00 00 04 00 00 00  |....j...........|
00000020  73 00 00 00 c7 03 00 00  04 01 00 00 00 08 00 00  |s...............|
00000c00  44 65 66 61 75 6c 74 20  46 6f 6c 64 65 72 20 20  |Default Folder  |
00000c10  00 40 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |.@..............|
```

These data is another confirmation of the suspicion on the shorts representing the used folder blocks. But this is something I already observed.

https://github.com/dagargo/emu3fs/blob/95b76a2ea8e38a68bbc224618000c7250a2d76e7/super.c#L111-L117

And a couple of final observations. 

* Every bank except the one in the EIV format, have the first short in the folder block as  `0x4000`. EIV has `0x8000`. Perhaps we could use this to identify the different FS.
* For every image (starting at `0x0` and reading 32 bits values), `data[8] = data[4] + data[5]`. Perhaps, this is the whole space reserved for the folders and its content.

**Comment by dagargo** [comment]
A few things.

> Every bank except the one in the EIV format, have the first short in the folder block as 0x4000. EIV has 0x8000. Perhaps we could use this to identify the different FS.

I said this but it's probably wrong because the endianness would be different and that's highly unlikely.

On another subject, I've noticed that, in some cases, the first directory block in the first directory is wrong. In the case of EMU3 devices this has no impact as the first directory content block is equivalent in concept as the first block of the first (and single) directory. The devices probably use the former while we must provide support for both. But, maybe, the filesystem itself has the limitation of having at least one directory and this way we could improve compatibility, Could you check this, @dromer? Just try to delete all folders in a disk and provide the output of `hexdump -C -v $image | head -n 3; b=0x$(hexdump -C -v $image | head -n 1 | awk '{print $10}'); ab=$(printf '%x' $((b * 512))); xa=0000$ab; hexdump -C -v $image | head -n 1000 | grep $xa -A 9`.

```
emu_std_v13.iso
00000000  45 4d 55 33 00 55 19 00  09 00 00 00 06 00 00 00  |EMU3.U..........|
00000010  0f 00 00 00 c0 00 00 00  02 00 00 00 07 00 00 00  |................|
00000020  cf 00 00 00 55 06 00 00  04 01 8a 55 19 00 00 00  |....U......U....|
00001200  44 65 73 69 67 6e 65 64  20 62 79 20 53 26 4d 2e  |Designed by S&M.|
00001210  00 40 09 00 10 00 11 00  12 00 ff ff ff ff ff ff  |.@..............|
```

As you can see, the first block should be `0x000f` but is `0x0009`. However, the next block is right (`0x0010`).

Also, I'd like to see the output of the same command above with an image with 5 directories because I'd like to know are those 2 bytes. 

Finally, let's not forget that directory content is indexed by blocks whereas file content is indexed by clusters, which is variable in size.

**Comment by dromer** [comment]
I'll see what I can do sometime this week.

**Comment by dagargo** [comment]
I've just pushed some commits to the `emu4` branch. The idea is to give support for first level directories to the module regardless of the device the image was created for.

It's a playground so I don't recommend you to use it at all, although it already lists both directory levels. If you want to give it a try do NOT write on the disk images because it's probably broken and you will corrupt them.

```
$ ls -la $EMU3_MOUNTPOINT
total 8
drwxr-xr-x  2 root root  512 ago 16 16:33  .
drwxr-x---+ 5 root root 4096 feb 14  2021  ..
drwxr-xr-x  2 user user    0 ago 16 16:33 'Default Folder'
drwxr-xr-x  2 user user 3584 ago 16 16:33 'New Folder'

% ls -lia $EMU3_MOUNTPOINT/Default\ Folder
total 32
128 drwxr-xr-x 2 user user     0 ago 16 16:33  .
  1 drwxr-xr-x 2 root root   512 ago 16 16:33  ..
144 -rw-r--r-- 1 user user 10738 ago 16 16:33 'Untitled Bank'
144 -rw-r--r-- 1 user user 10738 ago 16 16:33 'Untitled Bank'
144 -rw-r--r-- 1 user user 10738 ago 16 16:33 'Untitled Bank'
```

Notice that whenever a directory entry is repeated this is what happens. All directory entries with the same name point to the same inode. This is NOT gonna change but this only happen because devices allow it; emu3fs doesn't allow this.

In summary, these are the things to remember.

* The root directory directory can ony contain directories.
* Each of these can only contain files.
* In order to work in the actual devices, every disk should have at least one directory in which the banks will be stored.
* The remaining files and folders would be found only by EIV series.

I'd like to know if it is allowed in the EIV devices to delete all the folders. If you could check this would be awesome.

While devices only allow you to write 100 banks, I'm thinking of giving support the maximum physically allowed (128). What do you think?

**Comment by raybellis** [comment]
> Notice that whenever a directory entry is repeated this is what happens. All directory entries with the same name point to the same inode. This is NOT gonna change but this only happen because devices allow it; emu3fs doesn't allow this.

Would it make sense to encode the bank number in the virtual file name, e.g. `B001 - Untitled Bank` to mirror what EOS does?

**Comment by dagargo** [comment]
Many years ago I had the same thought and, IIRC, I implemented exactly that. But I remove this feature.

The reason is that we are working with storage devices that need to be compatible across different samplers and computers. If we prepend the bank number to the bank name at the VFS layer, the user will see the name as `B001 - Untitled Bank` that is nice, no problem so far. But, what happens when the user creates a new bank (or folder)? There are 2 possibilities here.

* emu3fs prepends the `Bxxx - ` at the VFS layer when a user creates a bank. In this case, it's very confusing because they could think that a bank with the name "A" would be just named "A" but they'll see "Bxxx - A" while what it's actually stored is jut "A". If you try to rename a bank is even worse because you should do something like `mv  "Bxxx - A" B`. Not to mention a few implementation intricacies.

* The user must consider this and do it manually, which makes no sense at all, because in this case they could choose a completely different name, besides that they don't know which ID must use.

Keep in mind that at the VFS layer there are many issues if the name is repeated. No solution is perfect and we'll always be affected by this. To avoid this, just create banks and folders with useful names.

In case it wasn't clear this is an old issue but there's nothing we can do about it.

But I'm open to suggestions and solutions.

**Comment by dagargo** [comment]
> While devices only allow you to write 100 banks, I'm thinking of giving support the maximum physically allowed (128). What do you think?

This was wrong.

```
(7 blocks / dir) * (16 entries / block) = 112 entries / dir
```

**Comment by dagargo** [comment]
I've just pushed some commits.

Every operation has been updated to reflect the new findings but the rename one, which now returns without doing anything. Let's test this first and solve the loose ends and then I'll continue.

It looks like it's working and all the tests pass.

There are a few things though.

* We need to address the bad blocks in the first directory. BTW, not only the first block is wrong in some cases. But as they only happen on single directory images is safe to assume that the first directory will use the first 7 blocks at most.

* The block list in a directory is never cleared up. Once a block has been assigned it will remain assigned to that directory. This is kind of easy to fix but I'm leaving this for later.

Regarding the bad blocks in the first directory, I'm proposing 2 solutions.

* Fix that at the mounting moment so the image will be fixed forever. This is the easiest and cleanest solution and can be tested easily. I'm voting for this one.

* Do not fix anything and use the expected values instead of the given ones. This keeps disk images pristine so it's safer in case we are wrong about this block values.

And, of course, it would be super nice if you could check this on the real devices as I only have an ESI-2000.

**Comment by dagargo** [comment]
I've just added the `emu4` filesystem. The only difference is the directory that is gonna be considered as the root directory. This hopefully eliminates the user fuss with the directories.

By default, ' emu3` is used, what does not break backwards compatibility.

Fixing the bad blocks is gonna be easier now.

**Comment by dromer** [comment]
@dagargo sorry I haven't been able to test this yet. Will try to do it this week, but no promises :#

**Comment by dagargo** [comment]
@dromer, don't be sorry. There is no deadline so no problem at all. :+1:

I'm still working on the few remaining loose ends but it's almost complete.

Once it's tested and all the necessary bugfixes are applied, I'll merge the `emu4` branch into the master.

**Comment by dagargo** [comment]
Just a small update.

* Rename operation has been completed so functionality is completed.
* I still have to do more testing so there will be bugs.
* In the end, I added a fix for the bad blocks in the first directory. I checked that it worked on my ESI-2000.

**Comment by dagargo** [comment]
@raybellis, I've just added support for extended attributes. The only one that is supported is the bank number. This is a much better alternative for the bank number in the name.

We can get it and set it only for files. As it's an attribute, be aware that changing a bank number *does not change other bank numbers*. If a bank number is repeated, my ESI-2000 shows the first one it finds in the directory for a given bank number.

This fixes a few weird things like the files with bank numbers greater than 99, whiches are the OS's judging by their name.

This is how it works.

``` 
$ sudo dd if=emu_std_v1.iso of=/dev/sda obs=4096
$ sudo mount -t emu4 /dev/sda /media/emu3
$ cd /media/emu3/Designed\ by\ S\&M./

$ getfattr -d -m ".*" * | head
# file: 12 String Guitar
user.bank.number="31"

# file: 4 Piece Horns 4M
user.bank.number="8"

# file: E3 Main Code
user.bank.number="109"
[...]

$ getfattr -n "user.bank.number" Textural\ Strings 
# file: Textural Strings
user.bank.number="6"

$ setfattr -n "user.bank.number" -v "99" Textural\ Strings 

$ getfattr -n "user.bank.number" Textural\ Strings 
# file: Textural Strings
user.bank.number="99"
```

With this and an alias or little bash script we can list the files with the bank number.

What do you think?

**Comment by dagargo** [comment]
I have been testing this thoroughly and seems to work nicely.

Some bugs and many improvements have been added since the last version and there is not much room left for improvements. It's a much better implementation now and feels like it's a finished product. I will not add new features but bugfixes.

I'll merge this branch manually whenever you have the time to test this. Again, no deadlines.

TIA.

**Comment by dagargo** [comment]
Any update on this?

**Comment by dromer** [comment]
Hey, I need to get back into this. Have not had any time to play with this stuff so will need to turn my head what needs to be done.
(not entirely off my radar and thanks for kicking the issue :) )

**Comment by dagargo** [comment]
As everything looks nice I've just merged the code from the branch `emu4` into the master to make the code available to others. I've done it manually to fix a couple of things and I've **deleted** the remote branch. So be careful with this.

However, I'm keeping this issue open because it still needs to be tested.

**Comment by dromer** [comment]
And I got reminded of this today over at https://github.com/izzyreal/vmpc-workspace/ testing (got stuck on linking to the akaidisk library hehe)

What do you want me to test exactly? Some concrete steps to reproduce would be helpful.

**Comment by dagargo** [comment]
We made a lot of assumptions, so I think we should test that everything works with the EIV, mainly the different folders as the remaining staff should work.

Perhaps, you could do something as this.

From the sampler
---
1. Format a new disk.
2. Create a folder called `A`.
3. Create a bank there with name `X`.

From the computer
---
1. Mount the disk as `emu4` and check that you can see both the directory and the file inside.
2. Create a new directory `B`.
3. Copy `X` to `B`.
4. Rename `X` bank to `Y`.
5. Remove the `A` directory and its content.

From the sampler again
---
1. Check that there is a new folder `B` with the bank `Y` and that bank loads properly.
2. Check that here is no `A` directory. 

It's pretty basic but covers reading, writing, creating and deleting directories, deleting files and renaming.

I'd test this but I don't have an EIV.

**Comment by dromer** [comment]
Sorry no time yet, but tomorrow I have a day off and will pull out the 'ol EMU and zip drive :)

**Comment by dromer** [comment]
Ok, went through all the motions (formatting the disk probably took the longest time) and totally can see a Folder `B` containing a bank `y`!
With folder `A` and bank `x` gone.

I'd say success? :)

**Comment by dromer** [comment]
Too bad we can't use emu3bm to manage the emu4 banks ;)

**Comment by dagargo** [comment]
> I'd say success? :)

I think it is! :tada: 

> Too bad we can't use emu3bm to manage the emu4 banks ;)

I've just managed to find an EIV disk image and I not only could mount it and see the content, I could do this too.

```
$ file /media/emu3/Symphoniks/Orchestralcolorz 
/media/david/emu3/Symphoniks/Orchestralcolorz: IFF data
```

So it looks like the format might be the [Interchange File Format](https://en.wikipedia.org/wiki/Interchange_File_Format). That could open editing the banks.

I've found [libiff](https://github.com/svanderburg/libiff), which is a library that could help us to do some some reverse engineering or even solve our thing completely but I haven't managed to compile it yet. I'll try to keep you informed of further research about this.

For now, let's close this issue and thank you very much for taking the time to test this.

In case of anything, open a new issue with the details.

**Comment by dromer** [comment]
Yes, I also noticed the bank is recognized as `IFF data`

**Comment by raybellis** [comment]
I've got various notes on the E4B format somewhere - I'll search them out and post here.   I know that I've decoded them sufficiently that my E4 disk browser at https://emu.tools/ can play the samples using WebAudio.    I plan to add the older E3 audio format at some point.

**Comment by raybellis** [comment]
I've pasted a JavaScript code sample that I use for decoding the E3S1 blobs within an E4B0 IFF file at https://gist.github.com/raybellis/5b08eaa99a4695498caf86812a10d15c

**Comment by dagargo** [comment]
Definitely, I'm gonna implement sample export. This is the TOC of a EIV bank.

```
$ emu4bm /media/emu4/Symphoniks/Don\'t\ Ever\ End 
Size: 10874750
FORM: 10874738
TOC1: 576
Size: 10874750
FORM: 10874738
TOC1: 576
E4Ma: Multimap
E4P1: The End of It 93
E4P1: My Ending     93
E4P1: SwellVintgeCrowd
E4P1: Uneedthe        
E4P1: Uneedthe E.     
E4P1: Do U Mindsweeper
E4P1: Single Osc Funk 
E4P1: Like an Avalansh
E3S1: Sample Clipboard
E3S1: My Ending     93
E3S1: SwellVintgeCrowd
E3S1: Uneedtha-5      
E3S1: Uneedtha+5      
E3S1: Uneedtha E.     
E3S1: Do U MindSweeper
E3S1: Single Osc Funk 
E3S1: Like an Avalansh
```

Probably, it will be implemented in a new project as `emu3bm` is quite messy due to the *containerless* nature of the EIII banks. To me, they look like a RAM memory dump or the like.

**Comment by dromer** [comment]
Happy to start engaging with `emu4bm` on this :)
Of course it would be cool if we could convert between these two formats at some point.
